gtk4-macros/src/composite_template_attr.rs

```diff
@@ -0,0 +1,196 @@
+use proc_macro2::{Ident, Span, TokenStream};
+use proc_macro_error::abort_call_site;
+use quote::quote;
+
+use std::env::current_dir;
+use std::path::Path;
+use std::string::ToString;
+
+use crate::ui_definitions_parser;
+use crate::ui_definitions_parser::UIDefinitions;
+use crate::util::*;
+
+fn read_template_file(filename: String) -> UIDefinitions {
+    if filename.is_empty() {
+        abort_call_site!(&r#"Usage: #[template("NAME")]"#);
+    }
+
+    match ui_definitions_parser::parse_file(&filename) {
+        Ok(template) => template,
+        Err(err) => match err {
+            ui_definitions_parser::Error::IOError(e) => {
+                let mut abs = filename.to_string();
+                let filename_path = Path::new(&filename);
+                if !filename_path.is_absolute() {
+                    if let Ok(working_dir) = current_dir() {
+                        let working_dir = working_dir.join(&filename_path);
+                        if let Some(working_dir) = working_dir.to_str() {
+                            abs = working_dir.to_string();
+                        }
+                    }
+                }
+                abort_call_site!("IO error when opening UI file {}: {}", abs, e);
+            }
+            ui_definitions_parser::Error::ParseError(e) => {
+                abort_call_site!("Error parsing UI XML file: {}", e);
+            }
+        },
+    }
+}
+
+/// Input is:
+/// struct MyGtkSubclass {
+/// }
+///
+/// We want to generate the struct members and the impl for
+/// * CompositeTemplate
+/// * ObjectSubclass
+/// * ObjectImpl
```

Comment:
As it generates `ObjectSubclass` and `ObjectImpl`  impls, is there a way to provide them manually or to somehow override (parts of) them? Do you have an idea how that could look like API-wise?

For example `ObjectImpl::get_property` / `set_property` would seem like something that needs to be overridden, and for those I have a couple of ideas how that can also be autogenerated to some degree (that's going to be part of https://github.com/gtk-rs/gtk-rs/pull/31). I'll add those ideas to a comment here tomorrow.

Reply:
Similarly `ObjectSubclass::class_init`, `new` and `ObjectImpl::constructed` would be things that will need a way to be overridden.

In short my idea was to not actually derive those impls (yet! that's the next step) but to have an auxiliary struct plus trait that handles this (in which case a derive macro would also be sufficient). Then you would call a single function on that as part of e.g. `ObjectSubclass::class_init` (and wherever else needed) to get it setup up properly.

Reply:
I see two possibilities: Add `#[attrs]` to modify what is automatically generated and what not or a separate `#[derive]` macro that generates those impls. "Unfortunately" derives are evaluated before attribute macros so I'm not sure if this is possible to do even though it would look less magical and would make more sense.

Reply:
The problem with `#[derive]` for exactly those traits however is that that derive macro would also need to compose with anything else in the class hierarchy (upwards and downwards) that might want to extend the impl

Reply:
I don't have a problem to remove those impls for now and look for a better solution later. Before the API gets finalized I think this deserves a second look though. Maybe it might be necessary to rename the template method to keep the simple `template` name for something more higher level than this?

Reply:
Yes if you don't have an idea right now let's keep it simple for now and a bit more verbose for users. But I agree that we should try finding a way to make this as convenient as the current version of the macro, while somehow still keeping it composable.

Maybe you could make this either
 - An attribute macro that adds the fields to the struct and derives the composite template trait
 - An attribute macro for another struct that adds the fields and derives some trait and then you'd have

```rust
#[template("path/to/ui")]
struct TemplateWidgets;

#[derive(CompositeTemplate)]
struct MyWidget {
    #[templates]
    widgets: TemplateWidgets,
    other_fields,
}
```

This would still require the `ObjectImpl` and `ObjectSubclass` impls as before but would at least autogenerate the struct from the UI file and autogenerate a function to bind them all.

Reply:
What would `#[templates]` do in that case? And where would I put the trait? Should probably not be in gtk4-macros but somewhere else right? I also don't know how the widget class would even know its parent class then: As the template is annotated one struct above there is no way to communicate here.

In this case we also can't ever statically check at any point in the future if the struct is actually derived from the parent class because only the widget struct would have access to the widget file.

We could also annotate the module. That might make it more flexible for any future stuff that would be done with the template file as it woukd be in scope and one could easily add an attribute to anything inside later to change the behavior of the macro etc.

```
#[template(file="path/to/ui")]
mod my_widget_imp {
    #[derive(CompositeTemplate)]
    struct MyWidget {
        #[templates]
        widgets: TemplateWidgets,
        other_fields,
    }
}
```

Reply:
I guess that could work too, yes. (annotating the module) Just means that you have to parse much more code and insert your things into possibly existing user code into the right places.

> What would #[templates] do in that case?

Just tell the derive macro that this is the thing that stores all the widgets and implements your new trait, so it can automatically map to register them all via the `CompositeTemplate` trait.

> And where would I put the trait? Should probably not be in gtk4-macros but somewhere else right?

You would have to put it into gtk4. Everything but the actual macro needs to be in normal crates.

>  I also don't know how the widget class would even know its parent class then: As the template is annotated one struct above there is no way to communicate here.

The user would (for now) still manually call into the `CompositeTemplate` trait functions from `class_init` and `constructed`, and that would dispatch to the autogenerated stuff.

> In this case we also can't ever statically check at any point in the future if the struct is actually derived from the parent class because only the widget struct would have access to the widget file.

Can you give an example of what you mean? Like the UI file says that this is a `GtkDialog` but actually the subclass in question is a `GtkApplicationWindow`?

That would indeed not be directly possible anymore, at least statically. At runtime it's as easy to check as otherwise.

Reply:
> An attribute macro for another struct that adds the fields and derives some trait and then you'd have

Changes like these would add an extra layer on the template usage end when accessing the template children :/ Also mind, not all children of a template are widgets - you also have things like gestures and menus.

Reply:
E.g. https://github.com/gtk-rs/gtk4-rs/pull/97/files#diff-a24e3fff03262f200f6a7bea3104da8ab03e6aa8234d005bf58b86e14e8829ebR83 feels a bit too long. You'd need to do this everywhere when accessing a template child.

Reply:
One Q: does this work with types defined within the application code?

When working with templates you'll often have something like this:

```xml
<template class="TemplateFoo" parent=GtkBox">
  <child>
    <object class="TemplateBar">
        <!-- properties, children, and gestures on TemplateBar here -->
    </object>
  </child>
</template>
```

where TemplateBar is another object/type defined in the application code.

Reply:
It should work as long as the type is in the namespace.
Concerning the separate widgets class: I could implement Deref. But that really just seems like a hack at that point.